File: routes/shutdown_container.go
package routes

import (
	"fmt"
	"os"
	"os/exec"
	"server/method"
	"server/method/db"
	"strings"

	"github.com/gofiber/fiber/v2"
)

func ShutdownContainer(app *fiber.App) {

	app.Get("/api/instance/shutdown/:bot_id/:admin_key", func(c *fiber.Ctx) error {

		var bot_id = c.Params("bot_id")
		var admin_key = c.Params("admin_key")

		if !method.ValidateAdminKey(admin_key) {
			fmt.Println("[Shutdown] Erreur admin_key", admin_key, " n'est pas valide!")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid admin_key!")
		}

		if _, err := os.Stat(method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id)); os.IsNotExist(err) {
			fmt.Println("[Shutdown] Erreur bot_id", bot_id, " n'existe pas!")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid bot_id!")
		}

		var IsOn = db.Get(bot_id + "_online")

		if IsOn {

			cliArray := []struct {
				L   string
				CWD string
			}{
				{
					L:   strings.Replace("pm2 stop {Code}", "{Code}", bot_id, 1),
					CWD: method.ProcessCWD(),
				},

				{
					L:   strings.Replace("pm2 delete {Code}", "{Code}", bot_id, 1),
					CWD: method.ProcessCWD(),
				},
			}

			for _, index := range cliArray {
				cmd := exec.Command("sh", "-c", index.L)
				cmd.Dir = index.CWD
				cmd.Stdout = os.Stdout
				cmd.Stderr = os.Stderr
				if err := cmd.Run(); err != nil {
					fmt.Print(err)
					return c.Status(fiber.StatusInternalServerError).SendString(err.Error())
				}
			}

			db.Set(bot_id+"_online", false)
		} else {
			fmt.Println("[Shutdown] Erreur tentative doublon!")
		}

		return c.SendStatus(200)
	})
}

File: src/Routes/v1/shutdown_container.ts
import { validateAdminKey } from '../../method/validateData.js';
import isContainerOn from '../../method/getContainerStatus.js';
import config from '../../method/getConfigData.js';
import { Request, Response } from 'express';
import { execSync } from "child_process";
import path from "node:path";
import fs from "node:fs";
import { db } from '../../method/database.js';
import { getOwnerByCode } from '../../method/getOwnerByCode.js';

export default {
    type: 'get',
    apiPath: '/api/v1/instance/shutdown/:bot_id/:force/:admin_key',
    run: async (req: Request, res: Response) => {
        const ownihrz_table = db.table("OWNIHRZ");
        const botId = req.params["bot_id"];
        const adminKey = req.params["admin_key"];
        const force = req.params["force"] === "true";

        try {
            if (!config?.api.apiToken) {
                console.log("Error: Failed to load config");
                return res.status(500).send("Failed to load config");
            }

            if (!validateAdminKey(adminKey)) {
                console.log("[Delete] Erreur admin_key n'est pas valide!");
                return res.status(403).send("Invalid admin_key!");
            }

            const botPath = path.join(process.cwd(), 'ownihrz', botId);
            if (!fs.existsSync(botPath)) {
                console.log("[Delete] Erreur bot_id n'existe pas!");
                return res.status(403).send("Invalid bot_id!");
            }

            if (await isContainerOn(botId)) {
                const shutdownCommands = [
                    { line: `pm2 stop ${botId} -f`, cwd: process.cwd() },
                    { line: `pm2 delete ${botId}`, cwd: process.cwd() }
                ];

                const shutdownPromises = shutdownCommands.map(async (cmd) => {
                    try {
                        execSync(cmd.line, { stdio: [0, 1, 2], cwd: cmd.cwd });
                        console.log(`Successfully executed: ${cmd.line}`);
                    } catch (e: any) {
                        console.error(`Error executing ${cmd.line}:`, e.toString().split('\n')[0]);
                    }
                });

                await Promise.all(shutdownPromises);
            } else {
                console.log(`[Shutdown] Container ${botId} already stopped`);
            }

            if (force) {
                const ownerid1 = getOwnerByCode(await ownihrz_table.get("CLUSTER"), botId);
                await ownihrz_table.set(`CLUSTER.${ownerid1}.${botId}.PowerOff`, true);
                console.log(`Force shutdown flag set for bot ${botId}`);
            }

            return res.sendStatus(200);
        } catch (error) {
            console.error(`Shutdown error for bot ${botId}:`, error);
            return res.status(500).send(`Shutdown failed for bot ${botId}`);
        }
    },
};